tests/ga/test_monitor.py

```diff
@@ -229,6 +277,158 @@ def test_failed_heartbeat_creates_telemetry(self, patch_report_heartbeat, _, *ar
+        self.assertEqual(0, patch_send_event.call_count)
+
+    @patch("azurelinuxagent.common.protocol.wire.WireClient.send_event")
+    @patch("azurelinuxagent.common.conf.get_lib_dir")
+    def test_collect_and_send_events_with_invalid_events(self, mock_lib_dir, patch_send_event, *args):
+        mock_lib_dir.return_value = self.lib_dir
+        dummy_events_dir = os.path.join(data_dir, "events", "collect_and_send_events_invalid_data")
+        fileutil.mkdir(self.event_logger.event_dir)
+
+        test_data = WireProtocolData(DATA_FILE)
+        monitor_handler, protocol = self._create_mock(test_data, *args)
+        monitor_handler.init_protocols()
+
+        for filename in os.listdir(dummy_events_dir):
+            shutil.copy(os.path.join(dummy_events_dir, filename), self.event_logger.event_dir)
+
+        monitor_handler.collect_and_send_events()
+
+        # Invalid events
+        self.assertEqual(0, patch_send_event.call_count)
+
+    @patch("azurelinuxagent.common.protocol.wire.WireClient.send_event")
+    @patch("azurelinuxagent.common.conf.get_lib_dir")
+    def test_collect_and_send_events_cannot_read_events(self, mock_lib_dir, patch_send_event, *args):
+        mock_lib_dir.return_value = self.lib_dir
+        dummy_events_dir = os.path.join(data_dir, "events", "collect_and_send_events_unreadable_data")
+        fileutil.mkdir(self.event_logger.event_dir)
+
+        test_data = WireProtocolData(DATA_FILE)
+        monitor_handler, protocol = self._create_mock(test_data, *args)
+        monitor_handler.init_protocols()
+
+        for filename in os.listdir(dummy_events_dir):
+            shutil.copy(os.path.join(dummy_events_dir, filename), self.event_logger.event_dir)
+            fileutil.chmod(os.path.join(self.event_logger.event_dir, filename), 111)
+
+        monitor_handler.collect_and_send_events()
+
+        # Invalid events
+        self.assertEqual(0, patch_send_event.call_count)
+
+    @patch("azurelinuxagent.common.conf.get_lib_dir")
+    def test_collect_and_send_generate_protocol_error_while_sending_events(self, mock_lib_dir, *args):
+        mock_lib_dir.return_value = self.lib_dir
+        fileutil.mkdir(self.event_logger.event_dir)
+
+        test_data = WireProtocolData(DATA_FILE)
+        monitor_handler, protocol = self._create_mock(test_data, *args)
+        monitor_handler.init_protocols()
+
+        sizes = [1, 2, 3]  # get the powers of 2, and multiple by 1024.
+
+        for power in sizes:
+            size = 2 ** power * 1024
+            self.event_logger.save_event(create_event_message(size))
+
+        with patch("azurelinuxagent.common.utils.restutil.http_post") as mock_http_post:
+            mock_http_post.return_value = ResponseMock(
+                status=restutil.httpclient.SERVICE_UNAVAILABLE,
+                response="")
+            monitor_handler.collect_and_send_events()
+
+        # This test validates that if we hit an issue while sending an event, we never send it again.
```

Comment:
> if we hit an issue while sending an event, we never send it again.

so a transitory issue may prevent us from sending telemetry? how hard would it be to improve that?

Reply:
Had an offline discussion. A new bug is created to investigate this issue.